engine/src/main/java/org/terasology/entitySystem/event/AbstractBeforeAfterEvent.java

```diff
@@ -0,0 +1,42 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.entitySystem.event;
+
+/**
+ * Immutable event to notify a change in a certain value.
+ * NOTE: This is a generic event and a more specific Event extending this event should be fired and reacted to.
+ */
+public class AbstractBeforeAfterEvent implements Event {
+    private final float oldValue;
+    private final float newValue;
+
+    /**
+     * Creates a new notification event on change in value.
+     */
+    AbstractBeforeAfterEvent(float oldValue, float newValue) {
+        this.oldValue = oldValue;
+        this.newValue = newValue;
+    }
+
+    /**
+     * Returns the value prior to change.
+     */
+    public float getOldValue() {
+        return oldValue;
+    }
+
+    /**
+     * Returns the value after change.
+     */
+    public float getNewValue() {
+        return newValue;
+    }
+
+    /**
+     * Returns the amount by which the value has been changed.
+     */
+    public float getChange() {
```

Comment:
If we want this convenience functionality you could introduce a 

```java
public abstract class FloatBeforeAfterEvent extends BeforeAfterEvent<Float> { ... }
```

for this purpose. 

For calculating the factor you could consider to return infinity in case the old value was zero, something like this (there's a difference between positive and negative infinity, though):

```java
    public float getFactor() {
        return oldValue != 0 ? newValue / oldValue : Float.POSITIVE_INFINITY;
    }
```

Reply:
I see, Again thanks for explaining!
But for this PR I don't think we should introduce FloatBeforeAfterEvent just for a getChange/Factor() method... 
The whole point of using generics I think is we don't need to have separate events for different datatypes.. 
So I think we can leave those separate functionalities to the specific events that extend this event.